akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +700,533 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+   *   is backpressured.
+   */
+  def statefulSink[T](partitioner: () ⇒ (ConsumerInfo, T) ⇒ Long, startAfterNbrOfConsumers: Int,
+                      bufferSize: Int = defaultBufferSize): Sink[T, Source[T, NotUsed]] =
+    Sink.fromGraph(new PartitionHub[T](partitioner, startAfterNbrOfConsumers, bufferSize))
+
+  /**
+   * Creates a [[Sink]] that receives elements from its upstream producer and routes them to a dynamic set
+   * of consumers. After the [[Sink]] returned by this method is materialized, it returns a [[Source]] as materialized
+   * value. This [[Source]] can be materialized an arbitrary number of times and each materialization will receive the
+   * elements from the original [[Sink]].
+   *
+   * Every new materialization of the [[Sink]] results in a new, independent hub, which materializes to its own
+   * [[Source]] for consuming the [[Sink]] of that materialization.
+   *
+   * If the original [[Sink]] is failed, then the failure is immediately propagated to all of its materialized
+   * [[Source]]s (possibly jumping over already buffered elements). If the original [[Sink]] is completed, then
+   * all corresponding [[Source]]s are completed. Both failure and normal completion is "remembered" and later
+   * materializations of the [[Source]] will see the same (failure or completion) state. [[Source]]s that are
+   * cancelled are simply removed from the dynamic set of consumers.
+   *
+   * This `sink` should be used when the routing function is stateless, e.g. based on a hashed value of the
+   * elements. Otherwise the [[#statefulSink]] can be used to implement more advanced routing logic.
+   *
+   * @param partitioner Function that decides where to route an element. The function takes two parameters;
+   *   the first is the number of active consumers and the second is the stream element. The function should
+   *   return the index of the selected consumer for the given element, i.e. int greater than or equal to 0
+   *   and less than number of consumers. E.g. `(size, elem) => math.abs(elem.hashCode) % size`.
+   * @param startAfterNbrOfConsumers Elements are buffered until this number of consumers have been connected.
+   *   This is only used initially when the stage is starting up, i.e. it is not honored when consumers have
+   *   been removed (canceled).
+   * @param bufferSize Total number of elements that can be buffered. If this buffer is full, the producer
+   *   is backpressured.
+   */
+  def sink[T](partitioner: (Int, T) ⇒ Int, startAfterNbrOfConsumers: Int,
+              bufferSize: Int = defaultBufferSize): Sink[T, Source[T, NotUsed]] =
+    statefulSink(() ⇒ (info, elem) ⇒ info.consumerIds(partitioner(info.size, elem)), startAfterNbrOfConsumers, bufferSize)
+
+  @DoNotInherit trait ConsumerInfo extends akka.stream.javadsl.PartitionHub.ConsumerInfo {
+
+    /**
+     * Identifiers of current consumers
+     */
+    def consumerIds: Array[Long]
```

Comment:
It seems the `consumerIds` is only materialized for the purpose of the partitioner. However, like in the case of `sink` not all use cases require an array here. Putting the array in the API means that we can never change the implementation later not to provide that array.

How about providing a method `def consumerIdByIdx(idx: Int): Long` and a method `def consumerIds: immutable.IndexedSeq[Long]` (that e.g. can be implemented by `ConsumerInfoImpl` directly). The reason is to provide both is the convience to be able to use all the methods of `IndexedSeq` quickly and still have a way to access an id without having to box the Long (which would be used in `sink` above).